### PR TITLE
feat(hal-core): initial sketch of hal-core APIs

### PR DESCRIPTION
**Pull request by hawkw**
This is still *very* rough. @iximeow, I would love to hear your
thoughts.

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

### CODE REVIEW

File: hal-core/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    type PAddr: Address;
+
+    /// The name of the architecture, as a string.
+    const NAME: &'static str;
+}
+
+pub trait Address:
+    Copy + ops::Add<usize> + ops::Sub<usize> + ops::AddAssign<usize> + ops::SubAssign<usize>
+{
+    // TODO(eliza): more ops (align up/down?)
```
[the remark this answers is not in the document]
Comment (hawkw) [reply hal-core/src/lib.rs]: yeah i was gonna add these later :)

File: hal-core/src/mem.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
+use crate::Address;
+use core::ops;
+
+/// A cross-platform representation of a memory region.
+#[derive(Debug, Clone)]
+pub struct Region<A: Address> {
+    base: A,
+    // TODO(eliza): should regions be stored as (start -> end) or as
```
[the remark this answers is not in the document]
Comment (hawkw) [reply hal-core/src/mem.rs]: yeah it doesn't _actually_ matter but i was assuming you would have an opinion

File: hal-core/src/boot.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+use crate::{mem, Architecture};
+use core::iter::Iterator;
+
+pub trait BootInfo {
+    type Arch: Architecture;
+    type MemoryMap: Iterator<Item = mem::Region<<Self::Arch as Architecture>::PAddr>>;
```
[the remark this answers is not in the document]
Comment (hawkw) [reply hal-core/src/boot.rs]: hmm, so the thought process was more that the interface that each architecture would have to implement would be fairly simple (e.g. iterator), we could then wrap that in a struct that adds lots more operations that can be defined cross-platform (e.g. "iterate through the memory map until you find a region that contains 0xBADADD4355" or whatever). and, iterator is something we don't need a heap for.

File: hal-core/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (hawkw) [reply hal-core/src/lib.rs]: fn offset maybe

File: hal-core/src/boot.rs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+    /// Returns the boot info's memory map.
+    fn memory_map(&self) -> Self::MemoryMap;
+
+    // TODO(eliza): figure out a non-bad way to represent boot command lines (is
```
[the remark this answers is not in the document]
Comment (hawkw) [reply hal-core/src/boot.rs]: they can have whatever protocol the bootloader emits, is the thing

File: hal-core/src/boot.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (iximeow) [reply hal-core/src/boot.rs]: oooh yeah that makes sense

**Comment by hawkw** [comment]
@iximeow added a bunch more `Address` methods :)

File: hal-core/src/boot.rs [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+pub trait BootInfo {
+    type Arch: Architecture;
+    type MemoryMap: Iterator<Item = mem::Region<Self::Arch>>;
+
+    /// Returns the boot info's memory map.
+    fn memory_map(&self) -> Self::MemoryMap;
+
+    // TODO(eliza): figure out a non-bad way to represent boot command lines (is
+    // it reasonable to convert them to rust strs when we barely have an operating
+    // system???)
```
Comment (hawkw) [inline hal-core/src/boot.rs]: Should we also add a fn to get a wide ptr to the kernel obj?

Comment (hawkw) [reply hal-core/src/boot.rs]: We can do this using the memory map iter

**Pull request merged**